Android/app/src/main/java/app/intra/net/doh/ServerConnectionFactory.java

```diff
@@ -52,13 +57,28 @@ public static String getIpString(Context context, String url) {
     Resources res = context.getResources();
     String[] urls = res.getStringArray(R.array.urls);
     String[] ips = res.getStringArray(R.array.ips);
+    String ret = "";
     for (int i = 0; i < urls.length; ++i) {
       // TODO: Consider relaxing this equality condition to a match on just the domain.
       if (urls[i].equals(url)) {
-        return ips[i];
+        ret = ips[i];
+        break;
+      }
+    }
+
+    try {
+      String domain = new URL(url).getHost();
+      String extraIPs = RemoteConfig.getExtraIPs(domain);
+      if (ret.isEmpty()) {
+        ret = extraIPs;
+      } else if (!extraIPs.isEmpty()) {
+        ret += "," + extraIPs;
       }
+    } catch (MalformedURLException e) {
+      LogWrapper.logException(e);
     }
-    return "";
+
+    return ret;
```

Comment:
Why does it return a String instead of a list? Is it for serialization with Go?

Reply:
Yes, and also because that's the native format that we read from servers.xml, because Android resource XML supports arrays of strings but not arrays of arrays of strings.